enhancements/insights/on-demand-data-gathering.md

```diff
@@ -0,0 +1,340 @@
+
+### Implementation Details/Notes/Constraints
+
+#### Current behavior
+The Insights operator currently runs and gathers the data in a single container. 
+Uploading of new archives, pruning old archives as well as downloading the last Insights analysis report (this is always per archive) is implemented in this one container/binary. 
+When there is a successful upload (to the console.redhat.com) of the archive then there is some short delay (waiting for the processing in console.redhat.com pipelines - see the [API documentation](https://console.redhat.com/docs/api/insights-results-aggregator/v2)) and then there is a request to get the last Insights analysis report per given cluster. 
+Finally the status section of the `insightsoperator.operator.openshift.io` CR is updated. Note that there is a retry logic in case of failed upload and after some time (and number of attempts) the Insights operator is marked as degraded.
+
+#### Updates required to allow on-demand data gathering
+The plan is to refactor the existing periodic data gathering to run in a separate Pod/Job. The existing Insights operator pod will still exist, but it would periodically do the following:
+- check the `insightsdatagather.config.openshift.io` CR to read current user configuration
+- create a new `datagather.insights.openshift.io` CR (in the `openshift-insights` namespace) based on the values from the above
+- spawn a new `job` to gather the Insights data, upload it to `console.redhat.com` and check that the archive was processed in console.redhat.com (this requires a new API endpoint to be implemented in https://console.redhat.com/docs/api/insights-results-aggregator/v2).
+  The new API endpoint will be identified by the cluster ID and will require `insightsRequestID` parameter. The endpoint will be queried with some delay (to allow processing)  
+- status attribute of the `datagather.insights.openshift.io` is updated within the job execution
+- wait for the job completion and if the processing part was successful then download the last Insights analysis report
+- check that the Insights analysis report ID is equal to the gathering job request ID. If the IDs are equal then update the corresponding job status with `AnalysisAvailable=True` condition and continue with the following step below, otherwise update the corresponding job status with `AnalysisAvailable=False` condition and do nothing 
+- based on the status of the last `datagather.insights.openshift.io` CR update the `insightsoperator.operator.openshift.io` CR and also the clusteroperator conditions
+- prune the old job definitions and `datagather.insights.openshift.io` CRs
+  
+As mentioned above the operator will work with CRs created only in the `openshift-insights` namespace and will create jobs only in this namespace. The periodically created jobs will have the `periodic` and current timestamp prefix. To start on-demand data gathering, the user must create a new CR with a unique name.
```

Comment:
Can we  use `generateName` for all cases and determine "when" based on creationTimestamp?

Reply:
What do you mean by `generateName` please? But yeah let's remove the timestamp prefix from the name and use the appropriate attribute instead.

Reply:
`generateName` is a property of the `metav1.ObjectMeta` object, if you specify a prefix in the `generateName` field and leave the `name` field empty, upon creation, the API server will generate a random 5 char string and append it to the end of the `generateName` field to get the `name` for the new object. This is how pods get named for example

```
metadata:
  generateName: my-kube-resource-
```
would become after being created
```
metadata:
  generateName: my-kube-resource-
  name: my-kube-resource-abcde
```
So this allows you to generate unique names for resources you're creating where the name doesn't matter all that much

Reply:
I see. Thank you. I will update it with `generateName`

Reply:
I have updated it here to include both proposals.